tests/modules/sms_test_gateway/src/Plugin/SmsGateway/Memory.php

```diff
@@ -107,30 +106,33 @@ public function incoming(SmsMessageInterface $sms_message) {
    * {@inheritdoc}
    */
   public function parseDeliveryReports(Request $request, Response $response) {
+    $gateway_id = $this->configuration['gateway_id'];
+    $memory_reports = \Drupal::state()->get('sms_test_gateway.memory.report', []);
+
     $data = Json::decode($request->request->get('delivery_report'));
-    $latest_reports = [];
+    $return = [];
     foreach ($data['reports'] as $report) {
-      $latest_reports[$report['message_id']] = new SmsDeliveryReport([
-        'recipient' => $report['recipient'],
-        'message_id' => $report['message_id'],
-        'time_sent' => $report['time_sent'],
-        'time_delivered' => $report['time_delivered'],
-        'status' => $report['status'],
-        'gateway_status' => $report['gateway_status'],
-        'gateway_status_code' => $report['gateway_status_code'],
-        'gateway_status_description' => $report['gateway_status_description'],
-      ]);
+      $message_id = $report['message_id'];
+      $new_report = (new SmsDeliveryReport())
+        ->setRecipient($report['recipient'])
+        ->setMessageId($message_id)
+        ->setStatus(SmsMessageReportStatus::DELIVERED)
+        ->setStatusMessage($report['status_message'])
+        ->setTimeQueued($report['time_sent'])
+        ->setTimeDelivered($report['time_delivered']);
+
+      // Set separately since this method should not have meaningful keys.
+      $return[] = $new_report;
+      // Reports in state must be keyed by message ID.
+      $memory_reports[$gateway_id][$message_id] = $new_report;
     }
 
-    // Update the latest delivery reports in \Drupal::state().
-    $gateway_id = $this->configuration['gateway_id'];
-    $reports = \Drupal::state()->get('sms_test_gateway.memory.report', []);
-    $reports[$gateway_id] = $latest_reports + $reports;
```

Comment:
+1 for these changes. There was actually a bug here which I found while testing some new gateways. This has fixed it.